ui/src/pages/Admin/AllStories/index.tsx

```diff
@@ -155,7 +256,10 @@ export const AllStories: React.FC = () => {
     if (allStories) {
       dispatch({ type: "INITIALIZE_AFTER_API", rows: allStories });
     }
-  }, [allStories]);
+    if (tagOptions) {
+      dispatch({ type: "INITIALIZE_AFTER_TAGS_API", rows: tagOptions });
+    }
+  }, [allStories, tagOptions]);
```

Comment:
ok to prevent this kinda stuff from happening, lets split up the data fetching from useSWR into it's own component.

```ts
export function AllStories() {
  const { data: tagOptions, error: tagError } = useSWR<string[]>("/api/tags");
  const { data: allStories, error } = useSWR<StoryView[] | undefined>(
    "/api/stories",
    fetchStories
  );

  // if tagError || error return error state;
  // if data undefined return loading state;

  return <AllStoriesInner tagOptions={tagOptions} allStories={allStories} />;
}

...

// tagOptions and allStories are always defined
function AllStoriesInner({ tagOptions, allStories }) {
  const [state, reducer] = useReducer(state, getInitState(tagOptions, allStories));
  // other hooks

  return JSX
}
```

Reply:
what is the advantage of doing this over our current way of using UseEffect?

Reply:
also if doing this, would i create 2 files? one for AllStories(), and one for AllStoriesInner, or just separate function in the current file

Reply:
> what is the advantage of doing this over our current way of using UseEffect?

This is just a cleaner pattern, I'd rather just instantiate state with correct data rather than dispatch later on.

Read up on **functional core, imperative shell**, which IMO is a really good design pattern to leverage - https://www.destroyallsoftware.com/screencasts/catalog/functional-core-imperative-shell. Here we have an imperative shell, which relies on network IO (side effect). This is abstracted away from the functional shell -> the `AllStoriesInner` component which is just responsible for rendering some UI. This is not exactly the case, as technically `useReducer` is a side effect, but thats a discussion for another day.

Here we are setting a clear separation of concerns here. Let's say someone wanted to add another data fetching request (another GET to a endpoint), they know they just have to edit the imperative component, `AllStories`. This would involve a new `useSWR` call, and then handling the according error and loading states. They don't have to worry about anything.

Think about testing as well, we can easily test either component for what it does. A more unit or UI regression test for the inner component, and a more integration based tested for the imperative wrapper component.

> also if doing this, would i create 2 files? one for AllStories(), and one for AllStoriesInner, or just separate function in the current file

Do whatever feels better to you lol

Reply:
i will give that a read. Another question tho. So right now in our handleSwitchChange function, we are using mutate. And after every mutate, the dispatch in our useeffect occurs i think which updates the state accordingly. Now with this new implementation, how would that change? Would mutate work fine??

Reply:
Yeah that would be fine, the props passed down would just change.

Reply:
Do you think we should do this code restructuring in this PR? Multiple people are working on this allstories stuff and might increase confusion and conflicts. It might be better to do this after functionality is complete. What do you think? I am fine doing it in this PR as well.

Reply:
Hmm, we can refactor afterwards if that works better for you.

Reply:
ok yeah lets do it after functionality is complete